### PR TITLE
`BlockId` removal: `runtime-api` refactor

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of:
- `ApiExt` methods:  `has_api`, `has_api_with`, `api_version`
- `CallApiAt` method: `runtime_version_at`
from: `BlockId<Block>` to: `Block::Hash`

It also changes the first argument of all generated **runtime API** calls from: `BlockId<Block>` to `Block::Hash`:
`runtime_api().function_call(&BlockId, ...)` -> `runtime_api().function_call(Hash, ...)`
   
This PR is part of BlockId::Number refactoring analysis (paritytech/substrate#11292)

polkadot companion: paritytech/polkadot#6721
cumulus companion: paritytech/cumulus#2190

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2327858

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2390771

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2394287

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2394288

File: client/consensus/babe/src/tests.rs [hunk 1]
```diff
@@ -45,7 +45,7 @@ use sp_keystore::{
 	SyncCryptoStore,
 };
 use sp_runtime::{
-	generic::{Digest, DigestItem},
+	generic::{BlockId, Digest, DigestItem},
```
Comment (bkchr) [inline client/consensus/babe/src/tests.rs:48]: Why this import?

File: client/service/src/client/client.rs [hunk 2]
```diff
@@ -1741,8 +1739,10 @@ where
 			.map_err(Into::into)
 	}
 
-	fn runtime_version_at(&self, at: &BlockId<Block>) -> Result<RuntimeVersion, sp_api::ApiError> {
-		let hash = self.backend.blockchain().expect_block_hash_from_id(at)?;
+	fn runtime_version_at(
+		&self,
+		hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/client.rs:1744]:
```suggestion
		hash: Block::Hash,
```

File: client/service/test/src/client/mod.rs [hunk 3]
```diff
@@ -291,20 +291,14 @@ fn client_initializes_from_genesis_ok() {
 	assert_eq!(
 		client
 			.runtime_api()
-			.balance_of(
-				&BlockId::Number(client.chain_info().best_number),
-				AccountKeyring::Alice.into(),
-			)
+			.balance_of(client.chain_info().best_hash, AccountKeyring::Alice.into(),)
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:294]:
```suggestion
			.balance_of(client.chain_info().best_hash, AccountKeyring::Alice.into())
```

File: client/service/test/src/client/mod.rs [hunk 4]
```diff
@@ -291,20 +291,14 @@ fn client_initializes_from_genesis_ok() {
...
 		1000
 	);
 	assert_eq!(
 		client
 			.runtime_api()
-			.balance_of(
-				&BlockId::Number(client.chain_info().best_number),
-				AccountKeyring::Ferdie.into(),
-			)
+			.balance_of(client.chain_info().best_hash, AccountKeyring::Ferdie.into(),)
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:301]:
```suggestion
			.balance_of(client.chain_info().best_hash, AccountKeyring::Ferdie.into())
```

File: client/service/test/src/client/mod.rs [hunk 5]
```diff
@@ -351,20 +345,14 @@ fn block_builder_works_with_transactions() {
 	assert_eq!(
 		client
 			.runtime_api()
-			.balance_of(
-				&BlockId::Number(client.chain_info().best_number),
-				AccountKeyring::Alice.into(),
-			)
+			.balance_of(client.chain_info().best_hash, AccountKeyring::Alice.into(),)
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:348]:
```suggestion
			.balance_of(client.chain_info().best_hash, AccountKeyring::Alice.into())
```

File: client/service/test/src/client/mod.rs [hunk 6]
```diff
@@ -351,20 +345,14 @@ fn block_builder_works_with_transactions() {
...
 		958
 	);
 	assert_eq!(
 		client
 			.runtime_api()
-			.balance_of(
-				&BlockId::Number(client.chain_info().best_number),
-				AccountKeyring::Ferdie.into(),
-			)
+			.balance_of(client.chain_info().best_hash, AccountKeyring::Ferdie.into(),)
```
Comment (bkchr) [inline client/service/test/src/client/mod.rs:355]:
```suggestion
			.balance_of(client.chain_info().best_hash, AccountKeyring::Ferdie.into())
```

File: primitives/api/src/lib.rs [hunk 7]
```diff
@@ -539,14 +537,17 @@ pub trait ApiExt<Block: BlockT> {
 		Self: Sized;
 
 	/// Checks if the given api is implemented and versions match.
-	fn has_api<A: RuntimeApiInfo + ?Sized>(&self, at: &BlockId<Block>) -> Result<bool, ApiError>
+	fn has_api<A: RuntimeApiInfo + ?Sized>(
+		&self,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline primitives/api/src/lib.rs:542]:
```suggestion
		at_hash: Block::Hash,
```

File: primitives/api/src/lib.rs [hunk 8]
```diff
@@ -539,14 +537,17 @@ pub trait ApiExt<Block: BlockT> {
...
+	) -> Result<bool, ApiError>
 	where
 		Self: Sized;
 
 	/// Check if the given api is implemented and the version passes a predicate.
 	fn has_api_with<A: RuntimeApiInfo + ?Sized, P: Fn(u32) -> bool>(
 		&self,
-		at: &BlockId<Block>,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline primitives/api/src/lib.rs:550]:
```suggestion
		at_hash: Block::Hash,
```

File: primitives/api/src/lib.rs [hunk 9]
```diff
@@ -555,7 +556,7 @@ pub trait ApiExt<Block: BlockT> {
 	/// Returns the version of the given api.
 	fn api_version<A: RuntimeApiInfo + ?Sized>(
 		&self,
-		at: &BlockId<Block>,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline primitives/api/src/lib.rs:559]:
```suggestion
		at_hash: Block::Hash,
```

File: primitives/api/src/lib.rs [hunk 10]
```diff
@@ -619,7 +620,10 @@ pub trait CallApiAt<Block: BlockT> {
 	) -> Result<Vec<u8>, ApiError>;
 
 	/// Returns the runtime version at the given block.
-	fn runtime_version_at(&self, at: &BlockId<Block>) -> Result<RuntimeVersion, ApiError>;
+	fn runtime_version_at(
+		&self,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline primitives/api/src/lib.rs:625]:
```suggestion
		at_hash: Block::Hash,
```

**Review by bkchr — APPROVED** [review]

File: client/consensus/babe/src/tests.rs (hunk 1 above)
Comment (michalkucharczyk) [reply client/consensus/babe/src/tests.rs:48]: > Why this import?

This import was removed from [`lib.rs`](https://github.com/paritytech/substrate/blob/a809327b0c7ace4e8757f2e65ed7b88a7c75b198/client/consensus/babe/src/lib.rs#L126). And this `mod tests` is still using  `BlockId`.

**Comment by michalkucharczyk** [comment]
bot fmt

File: primitives/api/test/tests/runtime_calls.rs [hunk 11]
```diff
@@ -57,9 +57,9 @@ fn calling_native_runtime_signature_changed_function() {
 		.set_execution_strategy(ExecutionStrategy::NativeWhenPossible)
 		.build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:60]:
```suggestion
	let best_hash = client.chain_info().best_hash;
```

File: primitives/api/test/tests/runtime_calls.rs [hunk 12]
```diff
@@ -36,9 +36,9 @@ use sp_consensus::SelectChain;
 fn calling_function_with_strat(strat: ExecutionStrategy) {
 	let client = TestClientBuilder::new().set_execution_strategy(strat).build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:39]:
```suggestion
	let best_hash = client.chain_info().best_hash;
```

File: primitives/api/test/tests/runtime_calls.rs [hunk 13]
```diff
@@ -68,27 +68,27 @@ fn calling_wasm_runtime_signature_changed_old_function() {
 		.set_execution_strategy(ExecutionStrategy::AlwaysWasm)
 		.build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:71]:
```suggestion
	let best_hash = client.chain_info().best_hash;
```

File: primitives/api/test/tests/runtime_calls.rs [hunk 14]
```diff
@@ -68,27 +68,27 @@ fn calling_wasm_runtime_signature_changed_old_function() {
...
 	assert_eq!(&res, &[1, 2]);
 }
 
 #[test]
 fn calling_with_both_strategy_and_fail_on_wasm_should_return_error() {
 	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::Both).build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
-	assert!(runtime_api.fail_on_wasm(&block_id).is_err());
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:82]:
```suggestion
	let best_hash = client.chain_info().best_hash;
```

File: primitives/api/test/tests/runtime_calls.rs [hunk 15]
```diff
@@ -68,27 +68,27 @@ fn calling_wasm_runtime_signature_changed_old_function() {
...
+	assert!(runtime_api.fail_on_wasm(at_hash).is_err());
 }
 
 #[test]
 fn calling_with_both_strategy_and_fail_on_native_should_work() {
 	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::Both).build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
-	assert_eq!(runtime_api.fail_on_native(&block_id).unwrap(), 1);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:90]:
```suggestion
	let best_hash = client.chain_info().best_hash;
```

File: primitives/api/test/tests/runtime_calls.rs [hunk 16]
```diff
@@ -97,8 +97,8 @@ fn calling_with_native_else_wasm_and_fail_on_wasm_should_work() {
 		.set_execution_strategy(ExecutionStrategy::NativeElseWasm)
 		.build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
-	assert_eq!(runtime_api.fail_on_wasm(&block_id).unwrap(), 1);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:100]: ditto. at_hash doesn't add any meaning imho

File: primitives/api/test/tests/runtime_calls.rs [hunk 17]
```diff
@@ -107,8 +107,8 @@ fn calling_with_native_else_wasm_and_fail_on_native_should_work() {
 		.set_execution_strategy(ExecutionStrategy::NativeElseWasm)
 		.build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
-	assert_eq!(runtime_api.fail_on_native(&block_id).unwrap(), 1);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:110]: ditto

File: primitives/api/test/tests/runtime_calls.rs [hunk 18]
```diff
@@ -117,18 +117,18 @@ fn use_trie_function() {
 		.set_execution_strategy(ExecutionStrategy::AlwaysWasm)
 		.build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
-	assert_eq!(runtime_api.use_trie(&block_id).unwrap(), 2);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:120]: ditto

File: primitives/api/test/tests/runtime_calls.rs [hunk 19]
```diff
@@ -117,18 +117,18 @@ fn use_trie_function() {
...
+	assert_eq!(runtime_api.use_trie(at_hash).unwrap(), 2);
 }
 
 #[test]
 fn initialize_block_works() {
 	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::Both).build();
 	let runtime_api = client.runtime_api();
-	let block_id = BlockId::Number(client.chain_info().best_number);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:128]: ditto

File: primitives/api/test/tests/runtime_calls.rs [hunk 20]
```diff
@@ -205,10 +205,10 @@ fn call_runtime_api_with_multiple_arguments() {
 	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::Both).build();
 
 	let data = vec![1, 2, 4, 5, 6, 7, 8, 8, 10, 12];
-	let block_id = BlockId::Number(client.chain_info().best_number);
+	let at_hash = client.chain_info().best_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:208]: ditto

File: primitives/api/test/tests/runtime_calls.rs [hunk 21]
```diff
@@ -225,8 +225,8 @@ fn disable_logging_works() {
 
 		let client = builder.build();
 		let runtime_api = client.runtime_api();
-		let block_id = BlockId::Number(0);
-		runtime_api.do_trace_log(&block_id).expect("Logging should not fail");
+		let at_hash = client.chain_info().genesis_hash;
```
Comment (melekes) [inline primitives/api/test/tests/runtime_calls.rs:228]:
```suggestion
		let genesis_hash = client.chain_info().genesis_hash;
```

**Review by melekes — APPROVED** [review]
👍 

just some minor comments about naming. e.g not sure why `at_hash` is better than `best_hash`

**Comment by michalkucharczyk** [comment]
> +1
> 
> just some minor comments about naming. e.g not sure why `at_hash` is better than `best_hash`

I don't know if better :). The `runtime.call` is executed _at_ given block.
If you believe it is unreadable or misleading I can change it (as your suggestions won't compile).

**Comment by melekes** [comment]
> I don't know if better :). The runtime.call is executed at given block.
If you believe it is unreadable or misleading I can change it (as your suggestions won't compile).

so my point is `best_hash` is better because it actually gives a reader an idea of what it is. If we use `at_hash`, then the reader is forced to look at `at_hash` definition because `at_hash` itself does not mean anything.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2404716

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2414111

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2414162

File: client/tracing/src/block/mod.rs [hunk 22]
```diff
@@ -229,7 +229,7 @@ where
 			.ok_or_else(|| Error::MissingBlockComponent("Extrinsics not found".to_string()))?;
 		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
 		let parent_hash = *header.parent_hash();
-		let parent_id = BlockId::Hash(parent_hash);
+		let parent_id: BlockId<Block> = BlockId::Hash(parent_hash);
```
Comment (acatangiu) [inline client/tracing/src/block/mod.rs:232]: nit:

you can remove this and just use `block_id_as_string(BlockId::<Block>::Hash(parent_hash))` in L301

**Review by acatangiu — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2416874

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2416873

**Comment by michalkucharczyk** [comment]
bot merge

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**